src/interaction/sessionrecording.cpp

```diff
@@ -0,0 +1,1207 @@
+    }
+
+    if (!foundCameraKeyframe)
+        signalPlaybackFinishedForComponent(recordedType::camera);
+}
+
+std::string SessionRecording::readHeaderElement(size_t readLen_chars) {
+    std::vector<char> readTemp(readLen_chars);
+    _playbackFile.read(&readTemp[0], readLen_chars);
+    return std::string(readTemp.begin(), readTemp.end());
+    //return std::string(&readTemp[0], &readTemp[readLen_chars - 1]);
+}
+
+void SessionRecording::signalPlaybackFinishedForComponent(recordedType type) {
+    if (type == recordedType::camera) {
+        _playbackActive_camera = false;
+        LINFO("Playback finished signal: camera");
+    }
+    else if (type == recordedType::time) {
+        _playbackActive_time = false;
+        LINFO("Playback finished signal: time");
+    }
+    else if (type == recordedType::script) {
+        _playbackActive_script = false;
+        LINFO("Playback finished signal: script");
+    }
+
+    if (!_playbackActive_camera && !_playbackActive_time && !_playbackActive_script ) {
+        _state = sessionState::idle;
+        _cleanupNeeded = true;
+        LINFO("Playback session finished");
+    }
+}
+
+void SessionRecording::stopPlayback() {
+    if (_state == sessionState::playback) {
+        _state = sessionState::idle;
+        _cleanupNeeded = true;
+        LINFO("Session playback stopped");
+    }
+}
+
+void SessionRecording::cleanUpPlayback() {
+    global::navigationHandler.stopPlayback();
+
+    Camera * camera = global::navigationHandler.camera();
+    ghoul_assert(camera != nullptr, "Camera must not be nullptr");
+    Scene * scene = camera->parent()->scene();
+    if (_timeline.size() != 0) {
+        unsigned int prevIdx = _timeline[_idxTimeline_cameraPtrPrev].idxIntoKeyframeTypeArray;
+        global::navigationHandler.setFocusNode(scene->sceneGraphNode(_keyframesCamera[prevIdx].focusNode));
+    }
+    global::scriptScheduler.stopPlayback();
+
+    _playbackFile.close();
+
+    //Clear all timelines and keyframes
+    _timeline.clear();
+    _keyframesCamera.clear();
+    _keyframesTime.clear();
+    _keyframesScript.clear();
+    _idxTimeline_nonCamera = 0;
+    _idxTime = 0;
+    _idxScript = 0;
+    _idxTimeline_cameraPtrNext = 0;
+    _idxTimeline_cameraPtrPrev = 0;
+    _hasHitEndOfCameraKeyframes = false;
+
+    _cleanupNeeded = false;
+}
+
+bool SessionRecording::isDataModeBinary() {
+    if (_recordingDataMode == recordDataMode::binary)
+        return true;
+    else
+        return false;
+}
+
+void SessionRecording::writeToFileBuffer(const double& src) {
+    const size_t writeSize_bytes = sizeof(double);
+    unsigned char const *p = reinterpret_cast<unsigned char const*>(&src);
+    memcpy((_keyframeBuffer + _bufferIndex), p, writeSize_bytes);
+    _bufferIndex += writeSize_bytes;
+}
+
+void SessionRecording::writeToFileBuffer(std::vector<char>& cvec) {
+    const size_t writeSize_bytes = cvec.size() * sizeof(char);
+    memcpy((_keyframeBuffer + _bufferIndex), &cvec[0], writeSize_bytes);
```

Comment:
Better:  `cvec.data()` instead of `&cvec[0]`

Reply:
Done